File: backend/app/api/routes/fees.py
import json
from datetime import date, datetime
from decimal import Decimal
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.responses import Response
from sqlalchemy import func, select
from sqlalchemy.orm import Session, joinedload

from app.core.audit import create_audit_log
from app.core.database import get_db
from app.core.dependencies import get_current_user, require_roles
from app.models.batch import Batch, StudentBatch
from app.models.enums import UserRole
from app.models.fee import FeePlan, Payment, StudentFee
from app.models.student import Student
from app.models.user import User
from app.schemas.common import PaginatedResponse
from app.schemas.fee import (
    DueItemResponse,
    FeePlanCreate,
    FeePlanResponse,
    FeePlanUpdate,
    PaymentCreate,
    PaymentResponse,
    StudentFeeCreate,
    StudentFeeResponse,
)
from app.services.fee_service import (
    calculate_due_amount,
    calculate_paid_amount,
    next_due_installment,
    parse_due_schedule,
    serialize_due_schedule,
)
from app.services.receipt_service import generate_receipt_pdf

router = APIRouter()


def _serialize_fee_plan(plan: FeePlan) -> FeePlanResponse:
    metadata = json.loads(plan.metadata_json) if plan.metadata_json else None
    return FeePlanResponse(
        id=plan.id,
        name=plan.name,
        type=plan.type,
        amount=plan.amount,
        metadata_json=metadata,
        created_at=plan.created_at,
    )


def _serialize_student_fee(student_fee: StudentFee) -> StudentFeeResponse:
    due_schedule = parse_due_schedule(student_fee.due_schedule_json)
    paid_amount = calculate_paid_amount(student_fee)
    due_amount = calculate_due_amount(student_fee)
    return StudentFeeResponse(
        id=student_fee.id,
        student_id=student_fee.student_id,
        batch_id=student_fee.batch_id,
        fee_plan_id=student_fee.fee_plan_id,
        total_fee=student_fee.total_fee,
        discount=student_fee.discount,
        due_schedule=due_schedule,
        paid_amount=paid_amount,
        due_amount=due_amount,
        created_at=student_fee.created_at,
    )


@router.post("/plans", response_model=FeePlanResponse, status_code=status.HTTP_201_CREATED)
def create_fee_plan(
    payload: FeePlanCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> FeePlanResponse:
    plan = FeePlan(
        institute_id=current_user.institute_id,
        name=payload.name,
        type=payload.type,
        amount=payload.amount,
        metadata_json=json.dumps(payload.metadata_json) if payload.metadata_json else None,
    )
    db.add(plan)
    db.commit()
    db.refresh(plan)
    return _serialize_fee_plan(plan)


@router.get("/plans", response_model=list[FeePlanResponse])
def list_fee_plans(
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> list[FeePlanResponse]:
    plans = list(
        db.scalars(
            select(FeePlan)
            .where(FeePlan.institute_id == current_user.institute_id)
            .order_by(FeePlan.created_at.desc())
        )
    )
    return [_serialize_fee_plan(plan) for plan in plans]


@router.patch("/plans/{plan_id}", response_model=FeePlanResponse)
def update_fee_plan(
    plan_id: int,
    payload: FeePlanUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> FeePlanResponse:
    plan = db.scalar(select(FeePlan).where(FeePlan.id == plan_id, FeePlan.institute_id == current_user.institute_id))
    if not plan:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Fee plan not found")

    updates = payload.model_dump(exclude_unset=True)
    if "metadata_json" in updates:
        updates["metadata_json"] = json.dumps(updates["metadata_json"]) if updates["metadata_json"] else None
    for key, value in updates.items():
        setattr(plan, key, value)
    db.commit()
    db.refresh(plan)
    return _serialize_fee_plan(plan)


@router.delete("/plans/{plan_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_fee_plan(
    plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN)),
) -> None:
    plan = db.scalar(select(FeePlan).where(FeePlan.id == plan_id, FeePlan.institute_id == current_user.institute_id))
    if not plan:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Fee plan not found")
    db.delete(plan)
    db.commit()
    return None


@router.patch("/batches/{batch_id}/plan", response_model=dict)
def assign_fee_plan_to_batch(
    batch_id: int,
    fee_plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> dict:
    batch = db.scalar(select(Batch).where(Batch.id == batch_id, Batch.institute_id == current_user.institute_id))
    if not batch:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Batch not found")
    plan = db.scalar(select(FeePlan).where(FeePlan.id == fee_plan_id, FeePlan.institute_id == current_user.institute_id))
    if not plan:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Fee plan not found")
    batch.fee_plan_id = fee_plan_id
    db.commit()
    return {"batch_id": batch_id, "fee_plan_id": fee_plan_id}


@router.post("/student-fees", response_model=StudentFeeResponse, status_code=status.HTTP_201_CREATED)
def create_student_fee(
    payload: StudentFeeCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> StudentFeeResponse:
    student = db.scalar(
        select(Student).where(Student.id == payload.student_id, Student.institute_id == current_user.institute_id)
    )
    if not student:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Student not found")
    batch = db.scalar(select(Batch).where(Batch.id == payload.batch_id, Batch.institute_id == current_user.institute_id))
    if not batch:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Batch not found")
    link = db.scalar(
        select(StudentBatch).where(
            StudentBatch.student_id == payload.student_id,
            StudentBatch.batch_id == payload.batch_id,
        )
    )
    if not link:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Student must be assigned to batch before fee mapping",
        )
    if payload.fee_plan_id:
        plan = db.scalar(
            select(FeePlan).where(FeePlan.id == payload.fee_plan_id, FeePlan.institute_id == current_user.institute_id)
        )
        if not plan:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Fee plan not found")

    existing = db.scalar(
        select(StudentFee).where(
            StudentFee.student_id == payload.student_id,
            StudentFee.batch_id == payload.batch_id,
            StudentFee.institute_id == current_user.institute_id,
        )
    )
    if existing:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Student fee mapping already exists")

    due_schedule = [
        {"due_date": item.due_date, "amount": str(item.amount)}
        for item in payload.due_schedule
    ]
    student_fee = StudentFee(
        institute_id=current_user.institute_id,
        student_id=payload.student_id,
        batch_id=payload.batch_id,
        fee_plan_id=payload.fee_plan_id,
        total_fee=payload.total_fee,
        discount=payload.discount,
        due_schedule_json=serialize_due_schedule(due_schedule),
    )
    db.add(student_fee)
    db.commit()
    db.refresh(student_fee)
    student_fee = db.scalar(
        select(StudentFee)
        .options(joinedload(StudentFee.payments))
        .where(StudentFee.id == student_fee.id)
    )
    return _serialize_student_fee(student_fee)


@router.get("/student-fees", response_model=PaginatedResponse)
def list_student_fees(
    student_id: int | None = Query(default=None),
    batch_id: int | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=200),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> PaginatedResponse:
    stmt = (
        select(StudentFee)
        .options(joinedload(StudentFee.payments))
        .where(StudentFee.institute_id == current_user.institute_id)
    )
    if student_id:
        stmt = stmt.where(StudentFee.student_id == student_id)
    if batch_id:
        stmt = stmt.where(StudentFee.batch_id == batch_id)

    if current_user.role == UserRole.STUDENT:
        if not current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Student profile missing")
        stmt = stmt.where(StudentFee.student_id == current_user.student_id)

    total = db.scalar(select(func.count()).select_from(stmt.subquery())) or 0
    rows = list(
        db.scalars(
            stmt.order_by(StudentFee.created_at.desc()).offset((page - 1) * page_size).limit(page_size)
        )
    )
    items = [_serialize_student_fee(item) for item in rows]
    return PaginatedResponse(total=total, page=page, page_size=page_size, items=items)


@router.post("/payments", response_model=PaymentResponse, status_code=status.HTTP_201_CREATED)
def create_payment(
    payload: PaymentCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> Payment:
    student_fee = db.scalar(
        select(StudentFee)
        .options(joinedload(StudentFee.payments))
        .where(StudentFee.id == payload.student_fee_id, StudentFee.institute_id == current_user.institute_id)
    )
    if not student_fee:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Student fee mapping not found")

    current_due = calculate_due_amount(student_fee)
    if Decimal(str(payload.amount)) > current_due:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Payment exceeds due amount")

    receipt_no = f"RCPT-{datetime.utcnow().strftime('%Y%m%d%H%M%S')}-{uuid4().hex[:6].upper()}"
    payment = Payment(
        institute_id=current_user.institute_id,
        student_fee_id=payload.student_fee_id,
        amount=payload.amount,
        paid_on=payload.paid_on,
        mode=payload.mode,
        receipt_no=receipt_no,
        remarks=payload.remarks,
        created_by=current_user.id,
    )
    db.add(payment)
    db.flush()
    create_audit_log(
        db=db,
        institute_id=current_user.institute_id,
        actor_user_id=current_user.id,
        action="FEE_PAYMENT_CREATED",
        entity="payment",
        entity_id=str(payment.id),
        before=None,
        after={
            "student_fee_id": payload.student_fee_id,
            "amount": str(payload.amount),
            "mode": payload.mode.value,
            "receipt_no": receipt_no,
        },
    )
    db.commit()
    db.refresh(payment)
    return payment


@router.get("/payments", response_model=PaginatedResponse)
def list_payments(
    student_fee_id: int | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=200),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> PaginatedResponse:
    stmt = (
        select(Payment)
        .join(StudentFee, StudentFee.id == Payment.student_fee_id)
        .where(Payment.institute_id == current_user.institute_id)
    )
    if student_fee_id:
        stmt = stmt.where(Payment.student_fee_id == student_fee_id)
    if current_user.role == UserRole.STUDENT:
        if not current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Student profile missing")
        stmt = stmt.where(StudentFee.student_id == current_user.student_id)
    total = db.scalar(select(func.count()).select_from(stmt.subquery())) or 0
    rows = list(
        db.scalars(
            stmt.order_by(Payment.created_at.desc()).offset((page - 1) * page_size).limit(page_size)
        )
    )
    return PaginatedResponse(
        total=total,
        page=page,
        page_size=page_size,
        items=[PaymentResponse.model_validate(item).model_dump(mode="json") for item in rows],
    )


@router.get("/payments/{payment_id}/receipt")
def download_receipt(
    payment_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Response:
    payment = db.scalar(
        select(Payment)
        .options(
            joinedload(Payment.student_fee).joinedload(StudentFee.student),
            joinedload(Payment.student_fee).joinedload(StudentFee.batch),
        )
        .where(Payment.id == payment_id, Payment.institute_id == current_user.institute_id)
    )
    if not payment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Payment not found")

    if current_user.role == UserRole.STUDENT:
        if not current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Student profile missing")
        if payment.student_fee.student_id != current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not allowed")

    pdf_bytes = generate_receipt_pdf(
        payment=payment,
        student_fee=payment.student_fee,
        student=payment.student_fee.student,
        batch=payment.student_fee.batch,
    )
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f"attachment; filename=receipt-{payment.receipt_no}.pdf"},
    )


@router.get("/dues", response_model=list[DueItemResponse])
def list_dues(
    batch_id: int | None = Query(default=None),
    student_id: int | None = Query(default=None),
    due_from: date | None = Query(default=None),
    due_to: date | None = Query(default=None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[DueItemResponse]:
    stmt = (
        select(StudentFee)
        .options(
            joinedload(StudentFee.student),
            joinedload(StudentFee.batch),
            joinedload(StudentFee.payments),
        )
        .where(StudentFee.institute_id == current_user.institute_id)
    )
    if batch_id:
        stmt = stmt.where(StudentFee.batch_id == batch_id)
    if student_id:
        stmt = stmt.where(StudentFee.student_id == student_id)
    if current_user.role == UserRole.STUDENT:
        if not current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Student profile missing")
        stmt = stmt.where(StudentFee.student_id == current_user.student_id)

    student_fees = list(db.scalars(stmt))
    results: list[DueItemResponse] = []
    for student_fee in student_fees:
        due_amount = calculate_due_amount(student_fee)
        if due_amount <= Decimal("0"):
            continue
        next_due_date, upcoming_amount = next_due_installment(student_fee)
        if due_from and next_due_date and date.fromisoformat(next_due_date) < due_from:
            continue
        if due_to and next_due_date and date.fromisoformat(next_due_date) > due_to:
            continue

        results.append(
            DueItemResponse(
                student_fee_id=student_fee.id,
                student_id=student_fee.student_id,
                student_name=student_fee.student.full_name,
                batch_id=student_fee.batch_id,
                batch_name=student_fee.batch.name,
                total_fee=student_fee.total_fee,
                discount=student_fee.discount,
                paid_amount=calculate_paid_amount(student_fee),
                due_amount=due_amount,
                next_due_date=next_due_date,
                upcoming_due_amount=upcoming_amount,
            )
        )
    return results

File: backend/src/routes/fees.js
const express = require("express");
const { randomUUID } = require("crypto");
const { connectDb } = require("../db");
const { requireAuth, requireRoles } = require("../auth");
const { parsePagination } = require("./helpers");
const { formatMoney, toMoneyNumber, sumPayments } = require("../utils/money");
const {
  createAuditLog,
  parseSchedule,
  getStudentFeeWithPayments,
  calculateDueAmount,
  nextDueInstallment,
  generateReceiptPdf
} = require("../services");

const router = express.Router();

function serializeStudentFee(row) {
  const paid = sumPayments(row.payments || []);
  const dueAmount = Math.max(toMoneyNumber(row.total_fee) - toMoneyNumber(row.discount) - paid, 0);
  return {
    id: row.id,
    student_id: row.student_id,
    batch_id: row.batch_id,
    fee_plan_id: row.fee_plan_id,
    total_fee: formatMoney(row.total_fee),
    discount: formatMoney(row.discount),
    due_schedule: parseSchedule(row.due_schedule_json),
    paid_amount: formatMoney(paid),
    due_amount: formatMoney(dueAmount),
    created_at: row.created_at
  };
}

router.post("/fees/plans", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const result = await db.run(
    `INSERT INTO fee_plans (institute_id, name, type, amount, metadata_json)
     VALUES (?, ?, ?, ?, ?)`,
    [
      req.user.institute_id,
      req.body.name,
      req.body.type,
      req.body.amount,
      req.body.metadata_json ? JSON.stringify(req.body.metadata_json) : null
    ]
  );
  const row = await db.get("SELECT * FROM fee_plans WHERE id = ?", [result.lastID]);
  row.metadata_json = row.metadata_json ? JSON.parse(row.metadata_json) : null;
  res.status(201).json(row);
});

router.get("/fees/plans", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const rows = await db.all("SELECT * FROM fee_plans WHERE institute_id = ? ORDER BY created_at DESC", [req.user.institute_id]);
  res.json(rows.map((row) => ({ ...row, metadata_json: row.metadata_json ? JSON.parse(row.metadata_json) : null })));
});

router.patch("/fees/plans/:planId", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const planId = Number(req.params.planId);
  const plan = await db.get("SELECT id FROM fee_plans WHERE id = ? AND institute_id = ?", [planId, req.user.institute_id]);
  if (!plan) {
    return res.status(404).json({ detail: "Fee plan not found" });
  }
  const updates = { ...req.body };
  if (Object.prototype.hasOwnProperty.call(updates, "metadata_json")) {
    updates.metadata_json = updates.metadata_json ? JSON.stringify(updates.metadata_json) : null;
  }
  Object.keys(updates).forEach((key) => updates[key] === undefined && delete updates[key]);
  if (Object.keys(updates).length) {
    const setSql = Object.keys(updates).map((key) => `${key} = ?`).join(", ");
    await db.run(`UPDATE fee_plans SET ${setSql}, updated_at = CURRENT_TIMESTAMP WHERE id = ?`, [
      ...Object.values(updates),
      planId
    ]);
  }
  const row = await db.get("SELECT * FROM fee_plans WHERE id = ?", [planId]);
  row.metadata_json = row.metadata_json ? JSON.parse(row.metadata_json) : null;
  res.json(row);
});

router.delete("/fees/plans/:planId", requireRoles("ADMIN"), async (req, res) => {
  const db = await connectDb();
  const planId = Number(req.params.planId);
  const row = await db.get("SELECT id FROM fee_plans WHERE id = ? AND institute_id = ?", [planId, req.user.institute_id]);
  if (!row) {
    return res.status(404).json({ detail: "Fee plan not found" });
  }
  await db.run("DELETE FROM fee_plans WHERE id = ?", [planId]);
  res.status(204).send();
});

router.patch("/fees/batches/:batchId/plan", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const batchId = Number(req.params.batchId);
  const feePlanId = Number(req.query.fee_plan_id || req.body.fee_plan_id);
  const batch = await db.get("SELECT id FROM batches WHERE id = ? AND institute_id = ?", [batchId, req.user.institute_id]);
  if (!batch) {
    return res.status(404).json({ detail: "Batch not found" });
  }
  const plan = await db.get("SELECT id FROM fee_plans WHERE id = ? AND institute_id = ?", [feePlanId, req.user.institute_id]);
  if (!plan) {
    return res.status(404).json({ detail: "Fee plan not found" });
  }
  await db.run("UPDATE batches SET fee_plan_id = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?", [feePlanId, batchId]);
  res.json({ batch_id: batchId, fee_plan_id: feePlanId });
});

router.post("/fees/student-fees", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const student = await db.get("SELECT id FROM students WHERE id = ? AND institute_id = ?", [req.body.student_id, req.user.institute_id]);
  if (!student) {
    return res.status(404).json({ detail: "Student not found" });
  }
  const batch = await db.get("SELECT id FROM batches WHERE id = ? AND institute_id = ?", [req.body.batch_id, req.user.institute_id]);
  if (!batch) {
    return res.status(404).json({ detail: "Batch not found" });
  }
  const link = await db.get("SELECT id FROM student_batches WHERE student_id = ? AND batch_id = ?", [req.body.student_id, req.body.batch_id]);
  if (!link) {
    return res.status(400).json({ detail: "Student must be assigned to batch before fee mapping" });
  }
  const existing = await db.get(
    "SELECT id FROM student_fees WHERE student_id = ? AND batch_id = ? AND institute_id = ?",
    [req.body.student_id, req.body.batch_id, req.user.institute_id]
  );
  if (existing) {
    return res.status(400).json({ detail: "Student fee mapping already exists" });
  }
  if (req.body.fee_plan_id) {
    const feePlan = await db.get("SELECT id FROM fee_plans WHERE id = ? AND institute_id = ?", [req.body.fee_plan_id, req.user.institute_id]);
    if (!feePlan) {
      return res.status(404).json({ detail: "Fee plan not found" });
    }
  }
  const dueSchedule = (req.body.due_schedule || []).map((item) => ({ due_date: item.due_date, amount: String(item.amount) }));
  const created = await db.run(
    `INSERT INTO student_fees (institute_id, student_id, batch_id, fee_plan_id, total_fee, discount, due_schedule_json)
     VALUES (?, ?, ?, ?, ?, ?, ?)`,
    [
      req.user.institute_id,
      req.body.student_id,
      req.body.batch_id,
      req.body.fee_plan_id || null,
      req.body.total_fee,
      req.body.discount || 0,
      JSON.stringify(dueSchedule)
    ]
  );
  const row = await getStudentFeeWithPayments(db, created.lastID);
  res.status(201).json(serializeStudentFee(row));
});

router.get("/fees/student-fees", requireAuth, async (req, res) => {
  const db = await connectDb();
  const { page, pageSize, offset } = parsePagination(req.query, 20);
  const params = [req.user.institute_id];
  let where = "WHERE institute_id = ?";
  if (req.query.student_id) {
    where += " AND student_id = ?";
    params.push(Number(req.query.student_id));
  }
  if (req.query.batch_id) {
    where += " AND batch_id = ?";
    params.push(Number(req.query.batch_id));
  }
  if (req.user.role === "STUDENT") {
    if (!req.user.student_id) {
      return res.status(403).json({ detail: "Student profile missing" });
    }
    where += " AND student_id = ?";
    params.push(req.user.student_id);
  }
  const total = await db.get(`SELECT COUNT(*) AS total FROM student_fees ${where}`, params);
  const rows = await db.all(
    `SELECT * FROM student_fees ${where}
     ORDER BY created_at DESC LIMIT ? OFFSET ?`,
    [...params, pageSize, offset]
  );
  const items = [];
  for (const row of rows) {
    items.push(serializeStudentFee(await getStudentFeeWithPayments(db, row.id)));
  }
  res.json({ total: total.total, page, page_size: pageSize, items });
});

router.post("/fees/payments", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const studentFee = await getStudentFeeWithPayments(db, req.body.student_fee_id);
  if (!studentFee || studentFee.institute_id !== req.user.institute_id) {
    return res.status(404).json({ detail: "Student fee mapping not found" });
  }
  const due = calculateDueAmount(studentFee);
  if (toMoneyNumber(req.body.amount) > due) {
    return res.status(400).json({ detail: "Payment exceeds due amount" });
  }
  const receiptNo = `RCPT-${new Date().toISOString().replace(/[-:TZ.]/g, "").slice(0, 14)}-${randomUUID().slice(0, 6).toUpperCase()}`;
  const created = await db.run(
    `INSERT INTO payments (institute_id, student_fee_id, amount, paid_on, mode, receipt_no, remarks, created_by)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      req.user.institute_id,
      req.body.student_fee_id,
      req.body.amount,
      req.body.paid_on,
      req.body.mode,
      receiptNo,
      req.body.remarks || null,
      req.user.id
    ]
  );
  await createAuditLog(db, {
    instituteId: req.user.institute_id,
    actorUserId: req.user.id,
    action: "FEE_PAYMENT_CREATED",
    entity: "payment",
    entityId: created.lastID,
    before: null,
    after: {
      student_fee_id: req.body.student_fee_id,
      amount: formatMoney(req.body.amount),
      mode: req.body.mode,
      receipt_no: receiptNo
    }
  });
  const payment = await db.get("SELECT * FROM payments WHERE id = ?", [created.lastID]);
  res.status(201).json(payment);
});

router.get("/fees/payments", requireAuth, async (req, res) => {
  const db = await connectDb();
  const { page, pageSize, offset } = parsePagination(req.query, 20);
  const params = [req.user.institute_id];
  let where = "WHERE p.institute_id = ?";
  if (req.query.student_fee_id) {
    where += " AND p.student_fee_id = ?";
    params.push(Number(req.query.student_fee_id));
  }
  if (req.user.role === "STUDENT") {
    if (!req.user.student_id) {
      return res.status(403).json({ detail: "Student profile missing" });
    }
    where += " AND sf.student_id = ?";
    params.push(req.user.student_id);
  }
  const total = await db.get(
    `SELECT COUNT(*) AS total FROM payments p JOIN student_fees sf ON sf.id = p.student_fee_id ${where}`,
    params
  );
  const items = await db.all(
    `SELECT p.* FROM payments p JOIN student_fees sf ON sf.id = p.student_fee_id ${where}
     ORDER BY p.created_at DESC LIMIT ? OFFSET ?`,
    [...params, pageSize, offset]
  );
  res.json({ total: total.total, page, page_size: pageSize, items });
});

router.get("/fees/payments/:paymentId/receipt", requireAuth, async (req, res) => {
  const db = await connectDb();
  const payment = await db.get("SELECT * FROM payments WHERE id = ? AND institute_id = ?", [
    Number(req.params.paymentId),
    req.user.institute_id
  ]);
  if (!payment) {
    return res.status(404).json({ detail: "Payment not found" });
  }
  const studentFee = await db.get("SELECT * FROM student_fees WHERE id = ?", [payment.student_fee_id]);
  const student = await db.get("SELECT * FROM students WHERE id = ?", [studentFee.student_id]);
  const batch = await db.get("SELECT * FROM batches WHERE id = ?", [studentFee.batch_id]);
  if (req.user.role === "STUDENT") {
    if (!req.user.student_id) {
      return res.status(403).json({ detail: "Student profile missing" });
    }
    if (studentFee.student_id !== req.user.student_id) {
      return res.status(403).json({ detail: "Not allowed" });
    }
  }
  const pdf = await generateReceiptPdf({ payment, studentFee, student, batch });
  res.setHeader("Content-Type", "application/pdf");
  res.setHeader("Content-Disposition", `attachment; filename=receipt-${payment.receipt_no}.pdf`);
  res.send(pdf);
});

router.get("/fees/dues", requireAuth, async (req, res) => {
  const db = await connectDb();
  const params = [req.user.institute_id];
  let where = "WHERE institute_id = ?";
  if (req.query.batch_id) {
    where += " AND batch_id = ?";
    params.push(Number(req.query.batch_id));
  }
  if (req.query.student_id) {
    where += " AND student_id = ?";
    params.push(Number(req.query.student_id));
  }
  if (req.user.role === "STUDENT") {
    if (!req.user.student_id) {
      return res.status(403).json({ detail: "Student profile missing" });
    }
    where += " AND student_id = ?";
    params.push(req.user.student_id);
  }
  const rows = await db.all(`SELECT * FROM student_fees ${where}`, params);
  const results = [];
  for (const row of rows) {
    const full = await getStudentFeeWithPayments(db, row.id);
    const dueAmount = calculateDueAmount(full);
    if (dueAmount <= 0) {
      continue;
    }
    const student = await db.get("SELECT full_name FROM students WHERE id = ?", [row.student_id]);
    const batch = await db.get("SELECT name FROM batches WHERE id = ?", [row.batch_id]);
    const { nextDueDate, upcomingAmount } = nextDueInstallment(full);
    if (req.query.due_from && nextDueDate && nextDueDate < req.query.due_from) {
      continue;
    }
    if (req.query.due_to && nextDueDate && nextDueDate > req.query.due_to) {
      continue;
    }
    results.push({
      student_fee_id: row.id,
      student_id: row.student_id,
      student_name: student?.full_name || "",
      batch_id: row.batch_id,
      batch_name: batch?.name || "",
      total_fee: formatMoney(row.total_fee),
      discount: formatMoney(row.discount),
      paid_amount: formatMoney(sumPayments(full.payments || [])),
      due_amount: formatMoney(dueAmount),
      next_due_date: nextDueDate,
      upcoming_due_amount: upcomingAmount == null ? null : formatMoney(upcomingAmount)
    });
  }
  res.json(results);
});

module.exports = router;
